### PR TITLE
Feature/manifests and blobs api for registries

### PR DESCRIPTION
**Pull request by ycliuhw**
We implemented two new registry APIs(manifests and blobs) for architecture fetching for OCI images (Because of this, we also migrated some of the registry providers from API v1 to v2 for public repositories using anonymous login).

## Checklist

 - [ ] ~Requires a [pylibjuju](https://github.com/juju/python-libjuju) change~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed~
 - [x] Comments answer the question of why design decisions were made

## QA steps

https://discourse.charmhub.io/t/advanced-config-to-host-your-own-juju-operator-images/5079

## Documentation changes

https://discourse.charmhub.io/t/advanced-config-to-host-your-own-juju-operator-images/5079

## Bug reference

https://bugs.launchpad.net/juju/+bug/1944801

### CODE REVIEW

File: docker/registry/image/image.go [hunk 1]
```diff
@@ -4,24 +4,62 @@
...
-	"github.com/juju/version/v2"
+	"encoding/json"
 
-	"github.com/juju/juju/tools"
+	"github.com/juju/errors"
+	"github.com/juju/version/v2"
 )
 
-type imageInfo struct {
+type ImageInfo struct {
```
Comment (wallyworld) [inline docker/registry/image/image.go:13]: should comment this

File: docker/registry/internal/base_client.go [hunk 2]
```diff
@@ -195,6 +189,10 @@ func (c baseClient) url(pathTemplate string, args ...interface{}) string {
 
 // Ping pings the baseClient endpoint.
 func (c baseClient) Ping() error {
+	if !c.repoDetails.IsPrivate() {
```
Comment (wallyworld) [inline docker/registry/internal/base_client.go:192]: It seems unclear why IsPrivate() is correlated with api version 1 or 2.
Should the APIVersion() method be used instead?

File: docker/auth.go [hunk 3]
```diff
@@ -167,6 +177,17 @@ func (rid ImageRepoDetails) String() string {
 	return string(d)
 }
 
+// Print returns a string for logging purpose.
+func (rid ImageRepoDetails) Print() string {
```
Comment (wallyworld) [inline docker/auth.go:181]: Should this be String() - ie for safety, the default behaviour should be to mask the tokens.
Then if people use "%s" they can't accidentally leak data.
If we need a method to print out the tokens, then that should be the exception, not the usual behaviour.

File: docker/registry/internal/acr.go [hunk 4]
```diff
@@ -43,19 +43,25 @@ func (c *azureContainerRegistry) WrapTransport(...TransportWrapper) error {
...
-		return c.fetchTags(url, &response)
-	}
-	if apiVersion == APIVersionV2 {
-		url := c.url("/%s/tags/list", imageName)
-		var response tagsResponseV2
-		return c.fetchTags(url, &response)
-	}
-	// This should never happen.
-	return nil, nil
+	url := c.url("/%s/tags/list", imageName)
```
Comment (wallyworld) [inline docker/registry/internal/acr.go:47]: So ACR is always api v2 now? Even for anonymous access?

**Review by wallyworld — APPROVED** [review]
To clarify - we no longer need api version 1 for ACR etc (given that v1 support was dropped for things like the Tags() method)? Is it that for ACR etc, we can always guarantee to use v2 for both public and private access? Do any of the repos still need v1? If so, can we drop APIVersion() etc? Or is v1 still needed sometimes?

File: docker/auth.go (hunk 3 above)
Comment (ycliuhw) [reply docker/auth.go:181]: Yeah, agreed. The usual behaviour be masking the tokens.

File: docker/registry/internal/acr.go (hunk 4 above)
Comment (ycliuhw) [reply docker/registry/internal/acr.go:47]: All the providers are using v2 now.

**Comment by ycliuhw** [comment]
> To clarify - we no longer need api version 1 for ACR etc (given that v1 support was dropped for things like the Tags() method)? Is it that for ACR etc, we can always guarantee to use v2 for both public and private access? Do any of the repos still need v1? If so, can we drop APIVersion() etc? Or is v1 still needed sometimes?



All the providers are using v2 now. As discussed, we keep the `APIVersion()` for future potential new API version implementations.

File: docker/registry/internal/base_client.go (hunk 2 above)
Comment (ycliuhw) [reply docker/registry/internal/base_client.go:192]: Ping here was used for credential validation which is why here skipped for non private. Also some of v1 root endpoint always requires credentials.
But now we always uses v2 which supports anonymous login. 
I am just removing this check here to allow Ping for both public and private cases.
There is one special case, gitlab root endpoint still does not work even for anonymous login. 
So I will skip ping for gitlab only.

**Comment by ycliuhw** [comment]
$$merge$$

File: docker/registry/internal/dockerhub.go [hunk 5]
```diff
@@ -48,6 +48,6 @@ func (c *dockerhub) DecideBaseURL() error {
 		addr += "/"
 	}
 	c.repoDetails.ServerAddress = addr
-	logger.Tracef("dockerhub repoDetails %#v", c.repoDetails)
+	logger.Tracef("dockerhub repoDetails %s", c.repoDetails)
```
Comment (hpidcock) [inline docker/registry/internal/dockerhub.go:51]: @ycliuhw repoDetails doesn't appear to implement the Stringer interface

**Pull request merged**